Hide color pickers in paragraph and button if no colors are available.

jorgefilipecosta opened the pull request:
If the theme sets the color palette to empty and disables custom colors, the color picker mechanism did not work correctly. It just showed a non-functional back color to choose from. Now we hide the color picker mechanism as in that case it is not possible to choose colors.

Closes: https://github.com/WordPress/gutenberg/issues/5447

## How Has This Been Tested?
Disable custom colors, and set the color pallete to empty. That can be done by adding the following lines to the functions.php of the theme:

```
add_theme_support( 'editor-color-palette' );
add_theme_support( 'disable-custom-colors' );

```

Verify in paragraph and button that the color panels are now hidden.
Enable some colors in the palette, and/or enable custom colors. Verify that if one or two the features are enabled the color panels are shown.

jorgefilipecosta committed:
Hide color pickers in paragraph and button if no colors are available.

If the theme sets the colors palette to empty and disable custom colors, the color picker mechanism did not work correctly. It just showed a non-functional back color to choose from. Now we hide the color picker mechanism as in that case it is not possible to choose colors.

aduth commented on blocks/conditional-panel-color/index.js:
```diff
@@ -0,0 +1,24 @@
+import { ifCondition, PanelColor, withContext } from '@wordpress/components';
+import { compose } from '@wordpress/element';
+
+export default compose(
```
Was trying to think of a way we could avoid duplicating the logic here between `ColorPalette` and `ConditionalPanelColor`. Unfortunate there's no easy way to avoid rendering a wrapper if none of its children render. I wonder if we could do a render prop which self-contains the null rendering:

```jsx
<EditorColors>
	{ ( { colors, disableCustomColors } ) => (
		<PanelColor title={ __( 'Text Color' ) } colorValue={ textColor }>
			<ColorPalette
				value={ textColor }
				onChange={ ( colorValue ) => setAttributes( { textColor: colorValue } ) }
				colors={ colors }
				disableCustomColors={ disableCustomColors }
			/>
		</PanelColor>
	) }
</EditorColors>
```

```js
const EditorColors = compose(
	withContext( 'editor' )(
		( settings, props ) => ( {
			colors: props.colors || settings.colors,
			disableCustomColors: props.disableCustomColors !== undefined ?
				props.disableCustomColors :
				settings.disableCustomColors,
		} )
	),
	ifCondition(
		( { colors, disableCustomColors } ) => ! ( isEmpty( colors ) && disableCustomColors )
	)
)( ( { children, ...props } ) => children( props ) );
```

This way, we don't need `ColorPalette` to retrieve color settings from context, because they'd always be passed in.

Thoughts?

aduth replied:
In either case, if we're creating a new component, it should be added to `blocks/index.js` to export on the global.

jorgefilipecosta replied:
Hi @aduth thank you for providing an alternative implementation, I really like the EditorColors component idea. I applied it to this PR.

youknowriad replied:
While I see the value of avoiding duplication, I think from a plugin author's perspective, duplication is fine or at least we could extract the logic into a `EditorColors` component but I think it should be baked in into these two components instead of using it at the block level. The block author could just use the `ColorPalette` or `ColorPanel` directly in that case.

jorgefilipecosta replied:
Hi @youknowriad, I understand your points. PanelColor is a generic component in @wordpress /components and we should not add the context information or conditional render directly to it. Maybe we should keep the context in the ColorPalette but change it to use a component that retrieves colors from the context. 
Then we create a new component HasColorsColorsToChoose or something similar that conditional render the children if colors are available or custom colors are enabled. And we use it to wrap color logic.
The other alternative is similar to what I had at the start a ConditionalPanelColor component that just renders PanelColor , but now we make use of a component that retrieves the colors from the state so no code duplication exists when comparing to the ColoPalette.

youknowriad replied:
>The other alternative is similar to what I had at the start a ConditionalPanelColor component that just renders PanelColor , but now we make use of a component that retrieves the colors from the state so no code duplication exists when comparing to the ColoPalette.

I was thinking more in that vein. The same we we have `wp.blocks.ColorPalette` we could have `we.blocks.ColorPanel` with context awareness. We can push the logic further by extracting the generic Color Palette to the components module so we'd also have `wp.components.ColorPalette` and `wp.components.ColorPanel` without any context awareness. Thoughs?

jorgefilipecosta replied:
I think it makes sense but maybe in the @blocks version, we should remove support for passing colors as props when that is a requirement the @components version can be used.
As if the behavior of the PanelColor in blocks is to not render if no colors are available if the plugin creator wants custom colors using the @blocks he would need to pass the colors also PanelColor which would only use them to check the conditional logic and does not need to know the colors for anything else.

youknowriad replied:
makes sense to me 👍 just thinking how we'd ensure backwards compatibility in that case. Maybe support it for a release or two with a deprecated message?

jorgefilipecosta replied:
Exactly, for the PanelColor @blocks  no deprecation is necessary as it is something new the existing one is at @components and will stay untouched. For the ColorPalette we would support but with a deprecation message when the prop is used saying to change to the @components version.

gziolo commented:
@jorgefilipecosta nice work. Can we go further and include `ContrastCheckerWithFallbackStyles` also in the `<EditColors />` wrapper as it will never render when there are other color components disabled?

This code looks so similar that it might make sense even go further and simplify it to:
```
<EditorColors
	setAttributes={ setAttributes }
	textColor={ textColor }
	backgroundColor={ backgroundColor }
	node={ this.nodeRef }
	isLargeText={ true }
	initialOpen={ false }
/>
```

One of the block uses `color` instead of `backgroundColor` so maybe it should be updated in that case to have this similar. It might be too much of the abstraction, but it might also make it easier to apply colors to other blocks :)

aduth commented on blocks/editor-colors/index.js:
```diff
@@ -0,0 +1,24 @@
+		} )
+	),
+	ifCondition(
+		( { colors, disableCustomColors } ) => ! ( isEmpty( colors ) && disableCustomColors )
```
I spent far too much time staring at this to make sense of it. Seems correct, not sure how we could make it clearer, but doing [De Morgan's](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) in my head a few times and thinking in inverses with disabling made this quite difficult to discern.

aduth reviewed (commented):
Not relevant to the changes here, but: What should be expected if I add this to my theme?

```php
add_theme_support( 'disable-custom-colors' );
add_theme_support( 'editor-color-palette', [] );
```

No idea where it's added, but this shows the color picker, with a single option for black. Doesn't seem correct to me (and even if it's working-as-intended, certainly not intuitive).

jorgefilipecosta commented:
Hi @gziolo I think it totally makes sense for the contrast checker to be inside the "conditional render", I missed it and I moved it inside.
Regarding an abstraction that includes the contrast verification, I'm thinking about it and maybe we should go for a component that abstracts the text color picker, the background color picker, and the contrast logic, to make block creation easier. For example in button, color logic occupies a significant percentage of the LOC's in the block. We will have to do a refactor to contrast logic (to remove a dependency on a CSS background rule) and during that refactor and after we land the class mechanism I think it would be a good idea to check what we can abstract. And create a component that does bigger abstraction.

I thought a little bit more and for now, I renamed EditorColors to ColorContext to make more evident that this is just a context retrieval component and not a UI mechanism. I also remove the conditional render from it and added a new prop hasColorsToChoose that blocks can make use to conditional render the children or for example show a warning or use predefined colors if colors are mandatory for the block to make sense.
I imagine that if we create a color abstraction component that includes text, background and contrast verification it will make use of this ColorContext Component to retrieve the color palette.


Hi @aduth,

>Not relevant to the changes here, but: What should be expected if I add this to my theme?

>add_theme_support( 'disable-custom-colors' );
add_theme_support( 'editor-color-palette', [] );
No idea where it's added, but this shows the color picker, with a single option for black. Doesn't seem correct to me (and even if it's working-as-intended, certainly not intuitive).

That happens, because we define the colors as arguments instead of as an array:
e.g:
add_theme_support( 'editor-color-palette',
    '#a156b4',
    '#d0a5db',
    '#eee',
    '#444'
);.
So when we use add_theme_support( 'editor-color-palette', [] ); The editor thinks we have one color the color "[]" and even sets the color attribute to "[]". The same happens if we set a color to any other invalid value e.g: add_theme_support( 'editor-color-palette', 1 );.
If we want to handle this problem I think the best way is to perform a validation step that checks if colors are valid and if not displays a warning in the console.
The PR https://github.com/WordPress/gutenberg/pull/5273 changes how themes should set the colors so as part of its backward compatibility we could add a validation to the colors that the theme sets and show a warning if they are not valid. The warning would handle the case add_theme_support( 'editor-color-palette', [] ); as in this case we are setting an invalid color.

Thank you both for reviews!

gziolo commented:
> Regarding an abstraction that includes the contrast verification, I'm thinking about it and maybe we should go for a component that abstracts the text color picker, the background color picker, and the contrast logic, to make block creation easier. For example in button, color logic occupies a significant percentage of the LOC's in the block. We will have to do a refactor to contrast logic (to remove a dependency on a CSS background rule) and during that refactor and after we land the class mechanism I think it would be a good idea to check what we can abstract. And create a component that does bigger abstraction.

Sounds great, but definitely not for this PR :)

gziolo commented on blocks/library/button/index.js:
```diff
@@ -101,24 +102,34 @@ class ButtonBlock extends Component {
-							isLargeText={ true }
-						/> }
+						<ColorContext>
+							{ ( { colors, disableCustomColors, hasColorsToChoose } ) => ( hasColorsToChoose ? [
```
It might make sense to refactor this code to work with `Fragment` component instead of an array:
```js
{ ( { colors, disableCustomColors, hasColorsToChoose } ) => ( hasColorsToChoose && (
    <Fragment>
        <PanelColor title={ __( 'Background Color' ) } colorValue={ color }>
        ....
    </Fragment>
) }
```

jorgefilipecosta replied:
I always question my self about what to use arrays or fragment component. For this cases, I will try to default to Fragment from now on. The code was updated.

gziolo commented:
It looks like this refactoring doesn't play well with CSS styles for the color notice:
![screen shot 2018-03-27 at 14 03 07](https://user-images.githubusercontent.com/699132/37966090-9d0cb40a-31c7-11e8-80f2-74ff66a7993d.png)

gziolo reviewed (changes requested):
ContrastCheckerWithFallbackStyles needs to be fixed.

jorgefilipecosta commented:
Hi @gziolo, I will found out about the problem with the contrast checker while testing this PR. Turns out it was not related to the changes here and also happened on the master (it was caused during a refactor to Notices). My PR that fixes it was already merged https://github.com/WordPress/gutenberg/pull/5800. I rebased this PR so this problem should now not happen.

youknowriad commented on blocks/color-palette/index.js:
```diff
@@ -77,12 +77,3 @@ export function ColorPalette( { colors, disableCustomColors = false, value, onCh
 	);
 }
-
-export default withContext( 'editor' )(
```
This is a breaking change in how this component works, you need to provide the colors as a prop now. Should we provide a BC alternative?

gziolo replied:
`ColorPalette` become a general purpose component after this refactoring. What we could do to provide backward compatibility is move `ColorPallete` to `components` module and still expose it in here with the context applied for the 2 upcoming release cycles. We would use `wp.components.ColorPallete` for the core blocks. What do you think?

youknowriad commented on blocks/library/button/index.js:
```diff
@@ -101,24 +102,36 @@ class ButtonBlock extends Component {
-							backgroundColor={ color }
-							isLargeText={ true }
-						/> }
+						<ColorContext>
```
Why did you choose to implement this as a component with children as function. I'm a little bit concerned about this because it's not so obvious for block authors? Do you think it's better to bake these checks inside `PanelColor` and `ColorPalette` instead?

jorgefilipecosta replied:
Hi @youknowriad, this pattern was used to avoid code duplication https://github.com/WordPress/gutenberg/pull/5570#discussion_r175208006.

gziolo replied:
@youknowriad, I would rather further consolidate those color panels as proposed in https://github.com/WordPress/gutenberg/pull/5570#issuecomment-375655807.

gziolo replied:
Hmm, this would work to:
```js
<PanelColor title={ __( 'Background Color' ) } attributeName="color" value={ color } />
```
😃

gziolo commented on blocks/panel-color/index.js:
```diff
@@ -0,0 +1,31 @@
+/**
+ * WordPress dependencies
+ */
+import { PanelColor } from '@wordpress/components';
```
It can be alsoimported as:
```js
import { PanelColor as PanelColorComponent } from '@wordpress/components';
```

Then we can name this component the same way as the file: `PanelColor`.

gziolo commented on blocks/panel-color/index.js:
```diff
@@ -0,0 +1,31 @@
+	);
+}
+
+export default withColorContext( ConditionalPanelColor );
```
To avoid `if ( ! hasColorsToChoose ) {` check it can be expressed as:
```js
export default compose(
    withColorContext,
    ifCondition( ( { hasColorsToChoose } ) => hasColorsToChoose ),
)( PanelColor );
```

Aside: `ifCondition` could support string to pass a prop name :)

gziolo commented on components/color-palette/index.js:
```diff
@@ -0,0 +1,79 @@
+		<div className="blocks-color-palette">
+			{ map( colors, ( color ) => {
+				const style = { color: color };
+				const className = classnames( 'blocks-color-palette__item', { 'is-active': value === color } );
```
Class names are no longer valid. They should be prefixed with `components-` not `blocks-` because they were moved to a different module.

gziolo commented on blocks/color-palette/index.js:
```diff
@@ -1,88 +1,11 @@
  */
-import { Dropdown, withContext } from '@wordpress/components';
-import { __, sprintf } from '@wordpress/i18n';
+import { ColorPalette } from '@wordpress/components';
```
Do we need to keep this file since we always use `PanelColor` in all core blocks?

jorgefilipecosta replied:
Yes, in order to be back compatible with blocks created by the community and tht use @blocks/ColorPalette.

gziolo reviewed (commented):
I had some more comments. It also needs to be rebased.

jorgefilipecosta commented:
Hi @gziolo, thank you for the catches, the feedback was addressed.

jorgefilipecosta commented:
This PR was rebased and all feedback was addressed.

gziolo reviewed (approved):
Let's ship it 🚢 
It had enough iterations, works as advertised and looks solid 👍
